Add u32checked_popcnt and u32unchecked_popcnt instructions

bergkvist opened the pull request:
## Describe your changes
~~Adds popcount implementation to `stdlib/asm/math/u32.masm`.~~

Add support for 2 new instructions in miden:
- `u32checked_popcnt` (36 cycles)
- `u32unchecked_popcnt` (33 cycles)

The implementation is based on this answer on StackOverflow:
https://stackoverflow.com/a/109025/8008396

```c
int numberOfSetBits(uint32_t i)
{
     // Java: use int, and use >>> instead of >>. Or use Integer.bitCount()
     // C or C++: use uint32_t
     i = i - ((i >> 1) & 0x55555555);        // add pairs of bits
     i = (i & 0x33333333) + ((i >> 2) & 0x33333333);  // quads
     i = (i + (i >> 4)) & 0x0F0F0F0F;        // groups of 8
     return (i * 0x01010101) >> 24;          // horizontal sum of bytes
}
```

<!--
## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.
-->

bobbinth commented:
Thank you! This is much better than what I had suggested in https://github.com/0xPolygonMiden/miden-vm/issues/616#issuecomment-1371934348!

I do think that rather than having this be a procedure in `stdlib` it would probably be better to add it as an assembly instruction (which will expand to the same set of operation). Would you be interested in implementing that?

bergkvist commented:
@bobbinth I've added the 2 instructions (`u32checked_popcnt`, `u32unchecked_popcnt`) now, based on your suggestion here: https://github.com/0xPolygonMiden/miden-vm/issues/616#issuecomment-1372662848, and removed the stdlib-version of popcount.

bobbinth commented on assembly/src/parsers/serde/mod.rs:
```diff
@@ -124,6 +124,8 @@ pub enum OpCode {
     U32UncheckedMin = 105,
     U32CheckedMax = 106,
     U32UncheckedMax = 107,
+    U32CheckedPopcnt = 229,
+    U32UncheckedPopcnt = 230,
```
I'd prefer to keep sequential numbering (i.e., change `U32CheckedPopcnt` to 108,  `U32UncheckedPopcnt` to 109 etc.). I know this will update many lines of code - but I think it is better long-term to do it like this.

bobbinth reviewed (changes requested):
Looks great! Thank you! I left one minor nit inline. But also, it would be great to:

1. Add the two new operation descriptions to the docs [here](https://github.com/0xPolygonMiden/miden-vm/blob/next/docs/src/user_docs/assembly/u32_operations.md?plain=1#L64).
2. Add a couple of tests for these operations somewhere [here](https://github.com/0xPolygonMiden/miden-vm/blob/next/miden/tests/integration/operations/u32_ops/bitwise_ops.rs).

bergkvist commented:
@bobbinth I've added docs + tests + proptests for `u32checked_popcnt` + `u32unchecked_popcnt` now. I also moved popcnt occurrences right below the other bitwise ops for consistency, and made the OpCode enum sequentially numbered.

bobbinth reviewed (approved):
All looks great! Thank you!

matthiasgoergens commented on assembly/src/parsers/serde/mod.rs:
```diff
@@ -108,156 +108,158 @@ pub enum OpCode {
-    U32UncheckedMin = 105,
-    U32CheckedMax = 106,
-    U32UncheckedMax = 107,
+    U32CheckedPopcnt = 92,
```
Just curious, why do we renumber everything?

bobbinth replied:
I think sequential numbering keeps it a bit cleaner - plus, it's not a lot of work to re-number :)
